src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,152 @@
+/* eslint-disable jsx-a11y/label-has-associated-control */
+import React, { useEffect, useRef, useState } from 'react';
+import cn from 'classnames';
+import { Todo } from '../../types/todo/Todo';
+
+type Props = {
+  todo: Todo;
+  isTodoLoading: (todoId: Todo['id']) => boolean;
+  isTempTodo?: boolean;
+  onTodoDelete?: (todoId: Todo['id']) => Promise<void>;
+  onTodoUpdate?: (
+    todoId: Todo['id'],
+    todoData: Partial<Pick<Todo, 'title' | 'completed'>>,
+  ) => Promise<void>;
+};
+
+export const TodoItem: React.FC<Props> = ({
+  todo,
+  isTodoLoading,
+  isTempTodo = false,
+  onTodoDelete,
+  onTodoUpdate,
+}) => {
+  const [isEditingTitleFormOpened, setIsEditingTitleFormOpened] =
+    useState(false);
+  const [editingTitle, setEditingTitle] = useState(todo.title);
+  const editingTitleInputRef = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    if (isEditingTitleFormOpened) {
+      setEditingTitle(() => todo.title);
+      editingTitleInputRef.current?.focus();
+    }
+  }, [isEditingTitleFormOpened, todo.title]);
+
+  const handleChangeTodoStatus = () => {
+    if (onTodoUpdate) {
+      onTodoUpdate(todo.id, { completed: !todo.completed });
+    }
+  };
+
+  const handleDeleteTodo = () => {
+    if (onTodoDelete) {
+      onTodoDelete(todo.id);
+    }
+  };
+
+  const handleUpdateTodoOrDeleteIfEditingTextIsEmpty = (
```

Comment:
Your function violates single responsibility principle. Create separate functions and move duplicated code to a new one